concordium-consensus/src/Concordium/GlobalState/Persistent/ReleaseSchedule.hs

```diff
@@ -0,0 +1,415 @@
+instance Applicative m => Cacheable m LegacyReleaseSchedule
+
+instance (MonadBlobStore m) => ReleaseScheduleOperations m (BufferedRef LegacyReleaseSchedule) where
+    type AccountRef (BufferedRef LegacyReleaseSchedule) = AccountAddress
+
+    addAccountRelease ts addr br = do
+        LegacyReleaseSchedule{..} <- refLoad br
+        refMake $!
+            LegacyReleaseSchedule
+                { rsFirstTimestamp = min rsFirstTimestamp ts,
+                  rsMap = Map.alter addAcc ts rsMap,
+                  rsEntryCount = 1 + rsEntryCount
+                }
+      where
+        addAcc Nothing = Just $! Set.singleton addr
+        addAcc (Just accs) = Just $! Set.insert addr accs
+
+    updateAccountRelease oldts ts addr br = do
+        LegacyReleaseSchedule{..} <- refLoad br
+        let remMap = Map.alter remAcc oldts rsMap
+        refMake $!
+            LegacyReleaseSchedule
+                { -- Since the new timestamp must be less than or equal to the old, this is correct.
+                  rsFirstTimestamp = min rsFirstTimestamp ts,
+                  rsMap = Map.alter addAcc ts remMap,
+                  rsEntryCount = rsEntryCount
+                }
+      where
+        remAcc Nothing = error "updateAccountRelease: no entry at expected release time"
+        remAcc (Just s) =
+            let s' = Set.delete addr s
+             in if Set.null s' then Nothing else Just s'
+        addAcc Nothing = Just (Set.singleton addr)
+        addAcc (Just accs) = Just $! Set.insert addr accs
+
+    processReleasesUntil ts br = do
+        rs@LegacyReleaseSchedule{..} <- refLoad br
+        if ts < rsFirstTimestamp
+            then return ([], br)
+            else do
+                let (newRS, accs) = go rs []
+                (accs,) <$> refMake newRS
+      where
+        go rs@LegacyReleaseSchedule{..} accum
+            | Map.null rsMap = (LegacyReleaseSchedule (Timestamp maxBound) Map.empty 0, accum)
+            | minTS <= ts =
+                go
+                    (rs{rsMap = newMap, rsEntryCount = rsEntryCount - fromIntegral (Set.size accs)})
+                    (accum ++ Set.toList accs)
+            | otherwise = (rs{rsFirstTimestamp = minTS}, accum)
+          where
+            -- Pattern match lazily in case the map is empty.
+            ~((minTS, accs), newMap) = Map.deleteFindMin rsMap
+
+-- |A set of accounts represented by 'AccountIndex'.
+newtype AccountSet = AccountSet {theAccountSet :: Set.Set AccountIndex}
+    deriving (Serialize)
+
+instance MonadBlobStore m => BlobStorable m AccountSet
+instance Applicative m => Cacheable m AccountSet
+
+-- |A release schedule for the P5 protocol version.  This uses a 'Trie.Trie' mapping 'Timestamp's
+-- to sets of accounts.
+data NewReleaseSchedule = NewReleaseSchedule
+    { -- |The first timestamp at which a release is scheduled (or the maximum possible timestamp if
+      -- no releases are scheduled).
+      rs5FirstTimestamp :: !Timestamp,
```

Comment:
Isn't it technically possible somebody schedules a release at that time?

From that perspective it would be better to use 0 as a sentinel value or really a Maybe, since that cannot be used by a valid transaction.

I see that would require a few changes down below but it does not seem that much would be affected.

Reply:
That's possible, and it's not a problem. Using a Maybe would just add complexity everywhere this is used. I've add documentation that you shouldn't infer that the map is empty because the timestamp is maxBound.